Issue244: Adding in Dynamic Proxy, Telemetry, and SimpleUI URLs

kb1lqc opened the pull request:
Per #244 this provides Proxy, Telemetry, and SimpleUI with the ability to use url's defined in the respective configuration files for Flask urls used to obtain telemetry and command units.

Todo:

- [x] Check that all configuration sample files have appropriate defaults
- [x] Catch expections when all new configuration elements fail to import (i.e. "REPLACEME" imported as an int.
- [ ] Test Proxy over internet
- [ ] Test Telemetry over internet
- [x] Consider configuration to disabled proxy POST operation for a first-attempt at somewhat safe hosting 

online and not allows stations to transmit data from anyone... licensed or not.

## Changelog
- Updated faradaybasicproxyio GET function with dynamic host address parameter
- Updated faradaybasicproxyio GET function with .format() string manipulation
- Updated faradaybasicproxyio POST function with dynamic host address parameter
- Updated faradaybasicproxyio POST function with .format() string manipulation
- Updated Proxy, Telemetry, and Simpleui to use `proxyhost` value or standard `host` in the flask section to set host of proxy server
- Updated javascript and HTML templates to import host/port information about proxy from Python
- Updated SimpleUI to check for command stations and catch exceptions when misconfigured.
- Added a very simplistict proxy flag to disabled adding anything to the POST queue.
- Updated proxy, telemetry, and simpleui to include relevant `disabletx`, `proxyhost` configuration values from command line

kb1lqc committed:
Initial code importing proxyHost value from config

Pull in proxyHost IP address from telemetry.ini. Also, manually set the IP
in proxy which needs to be updated to also pull in a configuration value.

kb1lqc committed:
Updated GET function with string formatting

Using the .format() functions to better organize the GET call URL.

kb1lqc committed:
Updated commens about flask server

Both telemetry and proxy flask server comments

kb1lqc committed:
Commented on simpleui and updated telemetry config

Updated the telemetery config truth file with proxyhost as well as added
logic to edit the value from command line. Also commented in simpleui
javascript on need to change hardcoding.

kb1lqc committed:
POST in simpleui using proxyhost in INI

Get proxy host address from simpleui.ini in the POST action for commands.
Got thrown off for a day with some wierd bug where nothing changed (any
edits) until I rebooted my computer... #()*&_)

kb1lqc committed:
Added POST exception for bad connection

When wrong IP or port is entered we ofted get a connectionerror exception.
This now is caught and dealt with safely on simpleui POST commands.

kb1lqc committed:
Dynamically getting javascript host/port

Dynamically get host and port values for proxy into the Javascript code
for commanding and data receiving.

kb1lqc committed:
Added dynamic return to webpage

Read in the flask host for simpleui to redirect dynamically.

kb1lqc committed:
Cleaned up pytest failures

kb1lqc committed:
Caught ValueError with simpleui configuration

Commanding with simpleui for local units can cause a valuerror if
'REPLACEME' is converted to an int. This catches the exception and returns
to the original simpleui without doing anything but printing an error
message.

kb1lqc committed:
Added really simple Proxy POST disable flag

A rudimentary attempt to disabled standard API transmissions from a
server. Likely used for simple web hosted Proxy instances. I make no
guarantee a even modestly skilled hacker can't get past this!

kb1lqc committed:
Added ValueError check for command configurations

If left unconfigured from initial state, the command local and remote
stations values can cause ValueError exceptions. This catches and responds
to them appropriately.

kb1lqc commented:
OK I tried to check sending Faraday proxy over an open internet connection to the world using a VPN. This did not apparently work. Maybe I did something wrong? @hdkmike have any ideas?

I essentially wanted to have flask server to "0.0.0.0" port 8000 which should set it to accept any connection over the internet (with debug=false set in Flask). I could connect over local router IP fine but the second I tried binding to a public IP nothing worked. I also checked my router port forwarding which should have been transferring all port 8000 activity to my desktop computer IP.

Overall I think I'm going to move on as this PR isn't specifically about serving over the internet, it's about cleaning up the IP/Port hard-coding. I was hoping this would be a two-for-one deal. Apparently not.

kb1lqc committed:
Added --disabletx option to proxy command line

kb1lqc commented:
@kb1lqd want me to pull this in without your review?

kb1lqd commented on faraday/proxy.py:
```diff
@@ -59,6 +59,7 @@
 parser.add_argument('--test-nodeid', dest='testnodeid', type=int, help='Set Faraday test mode nodeid')
 parser.add_argument('--test-rate', dest='testrate', default=1, type=int, help='Set Faraday test mode rate')
 parser.add_argument('--payload-size', dest='payloadsize', default=39, type=int, help='Set Faraday data mode payload size')
+parser.add_argument('--disabletx', type=int, default=0, help='Disable proxy from POSTing to Faraday thus preventing transmissions')
```
Not sure why you want this but not stopping.

kb1lqd reviewed (approved):
I ran the PR and proxy, telemetry, APRS, and simpleui worked as expected. Local commanding with simplyUI also worked, I didn't have the correct unit ID set but sending remote command caused the local unit to TX. I did not intentionally run alternate host addresses but saw that the default localhost worked. A quick review of the code looks like the updates are OK. 

I don't know why you're putting in the --disabletx option but sounds like you needed it and see a reason. I can see so that you ensure no transmissions but looks like you also block any local POST'ing that would NOT result in a transmission (i.e. local LED command).

kb1lqc committed:
Removed --disabletx

It wasn't working correctly and in all honesty let's just focus on it when
the need arises. Moving on.

kb1lqc commented:
@kb1lqd yeah that was rushed, I removed it in 0c5760f. Tested and all seems to be working. Next up on the whole project management tasks is unit testing. One day we can have nice things. One day :)
